### PR TITLE
feat(MultiSelectTag): Add MultiSelectTag widget

### PR DESCRIPTION
**Pull request by Peiren**
**Please check if the PR fulfills these requirements**

- [x] The commit(s) message(s) follows our [guidelines](https://github.com/Talend/react-talend-forms/blob/master/CONTRIBUTING.md#commit-message-format)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?**

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build / CI related changes
- [ ] Other... Please describe:

**What is the current behavior?** (You can also link to an open issue here)

**What is the new behavior?**

**Does this PR introduce a breaking change?**

- [ ] Yes
- [x] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:
JIRA: https://jira.talendforge.org/browse/TPSVC-1557
multiselecttag

### CODE REVIEW

File: src/widgets/MultiSelectTagWidget.js [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+				this.input.style.paddingLeft = '0px';
+				this.input.style.paddingTop = '0px';
+				this.input.style.height = '32px';
+			}
+		}
+	}
+
+	render() {
+		const { value, readonly, schema, options } = this.props;
+		const className = `dropdown ${this.state.showDropDownOptions ? 'open' : ''}`;
```
Comment (talend-glorieux) [inline src/widgets/MultiSelectTagWidget.js]: Please use classnames package for this kind of work.

File: src/widgets/MultiSelectTagWidget.js [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+		const dropDownOptions = this.getDropDownOptions();
+		const value2Label = mapValue2Label(options.enumOptions);
+		return (
+			<div className={className}>
+				{
+					!readonly && <input
```
Comment (talend-glorieux) [inline src/widgets/MultiSelectTagWidget.js]: Please avoid negative conditional statements. Especially if followed by an `&&`

File: src/widgets/MultiSelectTagWidget.js [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+				}
+				<div style={tagsWrapperStyle} ref={(tags) => this.adjustSearchInput(tags)}>
+					{
+						value.map((val, index) => {
+							const label = typeof val === 'object' ? val.label : value2Label[val];
+							return (
+								<Badge
+									label={label}
+									key={index}
+									onDelete={!readonly ? () => this.onRemoveTag(val) : undefined}
```
Comment (talend-glorieux) [inline src/widgets/MultiSelectTagWidget.js]: Please avoid the usage of undefined as empty/default value.

File: src/widgets/MultiSelectTagWidget.js [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+	maxWidth: '100%',
+};
+
+const menuItemStyle = {
+	overflowX: 'hidden',
+	textOverflow: 'ellipsis',
+	cursor: 'pointer',
+};
+
+const mapValue2Label = (enumOptions) => enumOptions.reduce((map, option) => {
```
Comment (talend-glorieux) [inline src/widgets/MultiSelectTagWidget.js]: Using arrow function isn't useful in this case. Please use regular functions.

File: src/widgets/MultiSelectTagWidget/MultiSelectTagWidget.js [hunk 5]
```diff
@@ -15,14 +15,13 @@ const INPUT_MIN_WIDTH = 135;
 
 function mapValue2Label(enumOptions) {
 	return enumOptions.reduce((map, option) => {
-		map[option.value] = option.label;
+		map[option.value] = option.label; // eslint-disable-line no-param-reassign
```
Comment (talend-glorieux) [inline src/widgets/MultiSelectTagWidget/MultiSelectTagWidget.js]: Please try and rewrite that line to remove the `eslint-disable-line` comment.

**Comment by jmfrancois** [comment]
Please put a screenshot of the widget as it is in the PR description

File: src/widgets/MultiSelectTagWidget/MultiSelectTagWidget.js (hunk 5 above)
Comment (Peiren) [inline src/widgets/MultiSelectTagWidget/MultiSelectTagWidget.js]: Though reassign map to another variable, still reference same object expect an extra variable name.
Here I think have to suppress this rule

File: src/widgets/MultiSelectTagWidget/MultiSelectTagWidget.scss [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+	.dropdown-toggle {
+		position: absolute;
+		right: 0;
+		top: 50%;
+		transform: translateY(-50%);
+		pointer-events: none
+	}
+
+	.dropdown-menu {
+		max-height: 200px;
```
Comment (frassinier) [inline src/widgets/MultiSelectTagWidget/MultiSelectTagWidget.scss]: With our theme `10px` = `1rem` so you could use `20rem`;

File: src/widgets/MultiSelectTagWidget/MultiSelectTagWidget.scss [hunk 7]
```diff
@@ -0,0 +1,42 @@
+.dropdown {
+	input.form-control {
+		padding-right: $padding-normal * 2 + $dropdown-caret-size;
+	}
+
+	.tags-container {
+		position: absolute;
+		padding-top: 2px;
+		top: 0;
+		max-width: calc(100% - #{$padding-normal * 2 + $dropdown-caret-size});
```
Comment (frassinier) [inline src/widgets/MultiSelectTagWidget/MultiSelectTagWidget.scss]: Pretty complex.

File: src/Form.js [hunk 8]
```diff
@@ -18,10 +20,11 @@ const TRIGGER_AFTER = 'after';
 const customWidgets = {
 	toggle: SwitchWidget,
 	tabs: TabsWidget,
+	multiSelectTag: MultiSelectTagWidget,
```
Comment (Peiren) [inline src/Form.js]: I hope to keep it multiSelectTag, multiSelect is a category includes checkboxes, multiSelectTag etc

File: src/widgets/MultiSelectTagWidget/MultiSelectTagWidget.js [hunk 9]
```diff
@@ -0,0 +1,267 @@
...
+				const { schema } = this.props;
+				if (schema.createIfNoneMatch) {
+					this.onCreateNewTag();
+				}
+			}
+			event.preventDefault();
+		}
+	}
+
+	onTagsMount(tags) {
```
Comment (Peiren) [inline src/widgets/MultiSelectTagWidget/MultiSelectTagWidget.js]: Not sure can do that with pure CSS

File: src/widgets/MultiSelectTagWidget/MultiSelectTagWidget.scss (hunk 7 above)
Comment (Peiren) [inline src/widgets/MultiSelectTagWidget/MultiSelectTagWidget.scss]: that's offsetWidth of the drop-down-caret element, padding on left and right plus width

File: package.json [hunk 10]
```diff
@@ -32,17 +33,24 @@
...
     "react-test-renderer": "15.4.0",
+    "rimraf": "^2.5.4",
+    "sass-loader": "^4.1.1",
     "semantic-release": "^4.3.5",
     "webpack": "^1.13.1"
   },
   "dependencies": {
-    "react-jsonschema-form": "^0.41.2"
+    "react-jsonschema-form": "^0.41.2",
+    "react-talend-components": "^0.50.0",
```
Comment (Peiren) [inline package.json]: I put react-talend-components in both peerDependencies and devDependencies.
Then put all required peerDependencies of react-talend-components in devDependencies.
You can check new package.json, hope that's the right way.

File: package.json [hunk 11]
```diff
@@ -32,17 +33,24 @@
...
+    "bootstrap-talend-theme": "^0.13.14",
```
Comment (Peiren) [inline package.json]: Not only for storybook, in multiSelectTagWidget.scss, we reference variable defined in bootstrap-talend-theme.

File: src/widgets/MultiSelectTagWidget/MultiSelectTagWidget.scss [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+		overflow-y: auto;
+		max-width: 100%;
+
+		li {
+			a[role='menuitem'] {
+				display: block;
+				overflow-x: hidden;
+				text-overflow: ellipsis;
+				cursor: pointer;
+			}
```
Comment (Peiren) [inline src/widgets/MultiSelectTagWidget/MultiSelectTagWidget.scss]: removed

**Pull request merged**